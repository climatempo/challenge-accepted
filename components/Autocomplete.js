import Autocomplete from 'vuejs-auto-complete'

let controller = null
let signal = null

// Precisei reescrever alguns métodos para resolver alguns bugs desse componente

Autocomplete.methods.request = function (url) {
  if (controller) {
    controller.abort()
  }
  controller = new AbortController()
  signal = controller.signal

  const promise = fetch(url, {
    method: this.method,
    credentials: this.getCredentials(),
    headers: this.getHeaders(),
    signal
  })
  return promise
    .then((response) => {
      if (response.ok) {
        this.error = null
        return response.json()
      }
      if (response.status === 404) {
        throw new Error('Resultado não encontrado.')
      } else {
        throw new Error('Resultado não encontrado.')
      }
    })
    .then((response) => {
      this.results = this.setResults(response)
      this.emitRequestResultEvent()
      this.loading = false
    })
    .catch((error) => {
      if (error.name !== 'AbortError') {
        this.error = error.message
        this.loading = false
      }
    })
}

Autocomplete.methods.select = function (obj) {
  if (!obj) {
    return
  }
  this.value = (this.resultsValue && obj[this.resultsValue]) ? obj[this.resultsValue] : obj.id
  this.display = this.formatDisplay(obj).replace('<b>', '').replace('</b>', '')
  this.selectedDisplay = this.display
  this.$emit('selected', {
    value: this.value,
    display: this.display,
    selectedObject: obj
  })
  this.$emit('input', this.value)
  this.close()
}

Autocomplete.methods.search = function (e) {
  e.preventDefault()
  if (e.type === 'input') {
    if (!this.display) {
      this.clear()
    }
    switch (true) {
      case typeof this.source === 'string':
        // No resource search with no input
        if (!this.display || this.display.length < 1) {
          this.$emit('clear')
          return
        }
        return this.resourceSearch(this.source + this.display)
      case typeof this.source === 'function':
        // No resource search with no input
        if (!this.display || this.display.length < 1) {
          return
        }
        return this.resourceSearch(this.source(this.display))
      case Array.isArray(this.source):
        return this.arrayLikeSearch()
      default:
        throw new TypeError('Type not defined')
    }
  }
}

Autocomplete.methods.blur = function () {
  if (controller) {
    controller.abort()
    this.loading = false
  }
  this.isFocussed = false
}

Autocomplete.methods.clear = function () {
  this.display = null
  this.value = null
  this.results = null
  this.error = null
  this.$emit('clear')
}

Autocomplete.methods.onlyClear = function () {
  this.display = null
  this.value = null
  this.results = null
  this.error = null
}

export default Autocomplete
