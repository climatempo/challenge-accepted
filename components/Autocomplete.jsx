import { useState, useEffect, useRef } from "react"
import debounce from "utils/useDebounce"
import styles from "styles/components/Autocomplete.module.scss"

const AutoComplete = ({ data, chooseCity, isOpen, setIsOpen }) => {
  const [suggestions, setSuggestions] = useState([])
  const [selected, setSelected] = useState(-1)
  const [displayValue, setDisplayValue] = useState("")
  const inputRef = useRef()

  // Observa abertura do painel de busca e seta foco na barra
  useEffect(() => {
    isOpen ? inputRef.current.focus() : inputRef.current.blur()
  }, [isOpen])

  // Configura entrada de texto
  function handleChange(e) {
    // pega o texto digitado e guarda
    const query = e.target.value.toLowerCase()
    setDisplayValue(query)
  }

  // Pega o que é digitado, e quando o cliente para de digitar faz o debounce, e guarda na var debounced
  const debouncedValue = debounce(displayValue)

  // Observa mudanças na var debounced, e busca sugestões
  useEffect(() => {
    if (debouncedValue != "") {
      setSuggestions(findSuggestion(data, debouncedValue))
    } else {
      resetSelection()
    }
  }, [debouncedValue, data])

  // Função de busca de sugestões com a query digitada
  const findSuggestion = (data, query) => {
    const filterSuggestions = data.filter(
      (suggestion) =>
        suggestion
          .toLowerCase()
          .normalize("NFD")
          .replace(/\p{Diacritic}/gu, "")
          .indexOf(query) > -1
    )
    return filterSuggestions
  }

  function resetSelection() {
    setDisplayValue("")
    setSuggestions([])
    setSelected(-1)
  }

  function hideSearch() {
    resetSelection()
    setIsOpen(false)
  }

  // Configura navegação por teclado
  const handleKeyDown = (e) => {
    // Seta para cima
    if (e.keyCode === 38) {
      if (selected <= 0) {
        return
      }
      setSelected(selected - 1)
    }
    // Seta para baixo
    else if (e.keyCode === 40) {
      if (selected === suggestions.length - 1) {
        return
      }
      setSelected(selected + 1)
    }
    // Enter
    else if (e.keyCode === 13) {
      if (selected != -1) {
        chooseCity(selected)
      }
      hideSearch()
    } else if (e.keyCode === 27) {
      hideSearch()
    }
  }

  // Configura seleção por click
  function handleClick(e) {
    const index = e.target.getAttribute("index")
    console.log("index: ", index)

    // TODO: FIX BUG: Não seta a variável em mobile - em desk está funcionando
    setSelected(index)
    console.log("selected: ", selected)
    selected != -1 ? chooseCity(selected) : false
    hideSearch()
  }

  const Suggestions = () => {
    return (
      <ul className={styles.suggestions}>
        {suggestions.map((suggestion, index) => {
          return (
            <li
              className={`
                ${styles.item}
                ${index === selected ? styles.active : ""}
                `}
              key={suggestion}
              index={index}
              onClick={(e) => {
                handleClick(e)
              }}
              onMouseOver={(e) => {
                e.target.classList.add(`${styles.active}`)
                setSelected(e.target.getAttribute("index"))
              }}
              onMouseOut={(e) => {
                e.target.classList.remove(`${styles.active}`)
              }}
            >
              {suggestion}
            </li>
          )
        })}
      </ul>
    )
  }

  return (
    <div className={`${styles.wrapper} ${isOpen ? "" : styles.hidden}`}>
      <input
        type="search"
        value={displayValue}
        onFocus={resetSelection}
        onChange={handleChange}
        onKeyDown={handleKeyDown}
        className={styles.input}
        ref={inputRef}
      />
      <Suggestions />
    </div>
  )
}

export default AutoComplete
