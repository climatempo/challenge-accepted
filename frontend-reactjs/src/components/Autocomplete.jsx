import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';

const Container = styled.div`
  height: 57px;
  background-color: #fff;
  box-shadow: 0 0 5px 1px rgba(0, 0, 0, .1);
  flex-direction: row;
  justify-content: space-between;
  align-items: stretch;
  & input[type=search] {
    height: 100%;
    border: none;
    outline: none;
    padding: 0 10px;
    ::placeholder {
      font-size: 1.2em;
    }
  }
`;

export const INPUT_ID = 'ct_autocomplete_data';

const Autocomplete = ({
  list, value, onSearch, onChange,
}) => (
  <Container>
    <InputContainer
      list={list}
      value={value}
      onChange={onChange}
    />
    <SearchButton onSearch={onSearch} />
  </Container>
);

Autocomplete.propTypes = {
  list: PropTypes.array,
  value: PropTypes.string,
  onSearch: PropTypes.func,
  onChange: PropTypes.func,
};

export default Autocomplete;

const InputContainer = ({ list, value, onChange }) => (
  <div style={{ flexGrow: '1' }}>
    <input
      list={INPUT_ID}
      type="search"
      placeholder="Digite o nome da regiÃ£o"
      value={value}
      onChange={onChange}
    />
    <datalist id={INPUT_ID}>
      {list && list.map((l) => (
        <option key={l.id} value={l.name}>{l.name}</option>
      ))}
    </datalist>
  </div>
);

InputContainer.propTypes = {
  list: PropTypes.array,
  value: PropTypes.string,
  onChange: PropTypes.func,
};

const SearchButton = ({ onSearch }) => (
  <div
    style={{ padding: '15px' }}
    onClick={onSearch}
    role="presentation"
  >
    <img
      src={`${window.location.origin}/images/icons/search.png`}
      alt=""
      style={{ height: '23px' }}
    />
  </div>
);

SearchButton.propTypes = {
  onSearch: PropTypes.func,
};
