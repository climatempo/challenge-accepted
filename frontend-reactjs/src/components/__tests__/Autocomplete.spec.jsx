import React from 'react';
import toJson from 'enzyme-to-json';

import { mount } from '../../tests-setup';

import Autocomplete, { INPUT_ID } from '../Autocomplete';

const props = {
  list: [{ id: 1, name: 'foo' }, { id: 2, name: 'bar' }],
  value: 'foo',
  onChange: jest.fn(),
  onSearch: jest.fn(),
};

describe('Autocomplete Component', () => {
  let wrapper = null;

  beforeEach(() => {
    wrapper = mount(<Autocomplete {...props} />);
  });

  it('Should correctly render', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('Should exists', () => {
    expect(wrapper.find('input').exists()).toBeTruthy();
    expect(wrapper.find('datalist').exists()).toBeTruthy();
    expect(wrapper.find('SearchButton').exists()).toBeTruthy();
  });

  it('Should input and datalist is connected', () => {
    expect(wrapper.find('input').prop('list')).toBe(INPUT_ID);
    expect(wrapper.find('datalist').prop('id')).toBe(INPUT_ID);
  });

  it('Should showing datalist when inputting data', () => {
    wrapper.find('input').simulate('change', { target: { value: 'foo' } });
    expect(props.onChange).toHaveBeenCalled();
    expect(props.onChange.mock.calls[0][0].target.value).toBe('foo');
    expect(wrapper.find('datalist').children()).toHaveLength(2);
  });

});
