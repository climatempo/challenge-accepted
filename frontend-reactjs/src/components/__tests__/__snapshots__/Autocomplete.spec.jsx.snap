// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete Component Should correctly render 1`] = `
<Autocomplete
  list={
    Array [
      Object {
        "id": 1,
        "name": "foo",
      },
      Object {
        "id": 2,
        "name": "bar",
      },
    ]
  }
  onChange={[MockFunction]}
  onSearch={[MockFunction]}
  value="foo"
>
  <styled.div>
    <StyledComponent
      forwardedComponent={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "sc-bdVaJa",
            "isStatic": true,
            "lastClassName": "epvRUM",
            "rules": Array [
              "
  height: 57px;
  background-color: #fff;
  box-shadow: 0 0 5px 1px rgba(0, 0, 0, .1);
  flex-direction: row;
  justify-content: space-between;
  align-items: stretch;
  & input[type=search] {
    height: 100%;
    border: none;
    outline: none;
    padding: 0 10px;
    ::placeholder {
      font-size: 1.2em;
    }
  }
",
            ],
          },
          "displayName": "styled.div",
          "foldedComponentIds": Array [],
          "render": [Function],
          "styledComponentId": "sc-bdVaJa",
          "target": "div",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <div
        className="sc-bdVaJa epvRUM"
      >
        <InputContainer
          list={
            Array [
              Object {
                "id": 1,
                "name": "foo",
              },
              Object {
                "id": 2,
                "name": "bar",
              },
            ]
          }
          onChange={[MockFunction]}
          value="foo"
        >
          <div
            style={
              Object {
                "flexGrow": "1",
              }
            }
          >
            <input
              list="ct_autocomplete_data"
              onChange={[MockFunction]}
              placeholder="Digite o nome da regiÃ£o"
              type="search"
              value="foo"
            />
            <datalist
              id="ct_autocomplete_data"
            >
              <option
                key="1"
                value="foo"
              >
                foo
              </option>
              <option
                key="2"
                value="bar"
              >
                bar
              </option>
            </datalist>
          </div>
        </InputContainer>
        <SearchButton
          onSearch={[MockFunction]}
        >
          <div
            onClick={[MockFunction]}
            role="presentation"
            style={
              Object {
                "padding": "15px",
              }
            }
          >
            <img
              alt=""
              src="http://localhost/images/icons/search.png"
              style={
                Object {
                  "height": "23px",
                }
              }
            />
          </div>
        </SearchButton>
      </div>
    </StyledComponent>
  </styled.div>
</Autocomplete>
`;
